x/paychan/model.go

```diff
@@ -62,54 +61,13 @@ func (pc PaymentChannel) Copy() orm.CloneableData {
 	}
 }
 
-// PaymentChannelBucket is a wrapper over orm.Bucket that ensures that only
-// PaymentChannel entities can be persisted.
-type PaymentChannelBucket struct {
-	migration.Bucket
-	idSeq orm.Sequence
-}
-
 // NewPaymentChannelBucket returns a bucket for storing PaymentChannel state.
-func NewPaymentChannelBucket() PaymentChannelBucket {
-	b := migration.NewBucket("paychan", "paychan", orm.NewSimpleObj(nil, &PaymentChannel{}))
-	return PaymentChannelBucket{
-		Bucket: b,
-		idSeq:  b.Sequence("id"),
-	}
-}
-
-// Create adds given payment store entity to the store and returns the ID of
-// the newly inserted entity.
-func (b *PaymentChannelBucket) Create(db weave.KVStore, pc *PaymentChannel) (orm.Object, error) {
-	key, err := b.idSeq.NextVal(db)
-	if err != nil {
-		return nil, err
-	}
-	obj := orm.NewSimpleObj(key, pc)
-	return obj, b.Bucket.Save(db, obj)
+func NewPaymentChannelBucket() orm.ModelBucket {
+	b := newPaymentChannelObjectBucket()
+	return migration.NewModelBucket("paychan", orm.NewModelBucket(b))
 }
 
-// Save updates the state of given PaymentChannel entity in the store.
-func (b *PaymentChannelBucket) Save(db weave.KVStore, obj orm.Object) error {
-	if _, ok := obj.Value().(*PaymentChannel); !ok {
-		return errors.WithType(errors.ErrModel, obj.Value())
-	}
-	return b.Bucket.Save(db, obj)
-}
-
-// GetPaymentChannel returns a payment channel instance with given ID or
-// returns an error.
-func (b *PaymentChannelBucket) GetPaymentChannel(db weave.KVStore, paymentChannelID []byte) (*PaymentChannel, error) {
-	obj, err := b.Get(db, paymentChannelID)
-	if err != nil {
-		return nil, err
-	}
-	if obj == nil || obj.Value() == nil {
-		return nil, errors.Wrap(errors.ErrNotFound, "payment channel not found")
-	}
-	pc, ok := obj.Value().(*PaymentChannel)
-	if !ok {
-		return nil, errors.Wrap(errors.ErrNotFound, "payment channel not found")
-	}
-	return pc, nil
+func newPaymentChannelObjectBucket() orm.Bucket {
+	obj := orm.NewSimpleObj(nil, &PaymentChannel{})
```

Comment:
Nice API for me would be something like:

`return orm.NewModelBucket("paychan", &PaymentChannel{}).WithIndex(...)`

we could auto-create the obj for `orm.NewBucket` as well as use that model in `ModelBucket.Put` to enforce the type

Reply:
I address comments as I read them and now I see where this is going :speak_no_evil: 

Combining https://github.com/iov-one/weave/pull/615#discussion_r284143700 and https://github.com/iov-one/weave/pull/615#discussion_r283960138 I agree that changing the API to not explicitly require `orm.Bucket` instance is beneficial. I will work on this now.